test/ng/rootScopeSpec.js

```diff
@@ -2316,6 +2316,68 @@ describe('Scope', function() {
         }));
 
 
+        // See issue https://github.com/angular/angular.js/issues/16135
+        it('should deallocate the listener array entry', inject(function($rootScope) {
+          var remove1 = $rootScope.$on('abc', noop);
+          $rootScope.$on('abc', noop);
+
+          expect($rootScope.$$listeners['abc'].length).toBe(2);
+          expect(0 in $rootScope.$$listeners['abc']).toBe(true);
+
+          remove1();
+
+          expect($rootScope.$$listeners['abc'].length).toBe(2);
+          expect(0 in $rootScope.$$listeners['abc']).toBe(false);
+        }));
+
+
+        it('should call next listener when removing current', inject(function($rootScope) {
+          var remove1 = $rootScope.$on('abc', function() { remove1() });
+
+          var listener2 = jasmine.createSpy();
+          var remove2 = $rootScope.$on('abc', listener2);
+
+          var listener3 = jasmine.createSpy();
+          var remove3 = $rootScope.$on('abc', listener3);
+
+          $rootScope.$broadcast('abc');
+          expect(listener2).toHaveBeenCalled();
+          expect(listener3).toHaveBeenCalled();
+        }));
```

Comment:
You could run a second round (it this and the subsequent tests) to make sure the listener will be called correctly and nothing will break after the initial run. E.g.:

```js
listener2.calls.reset();
listener3.calls.reset();

$rootScope.$broadcast('abc');
expect(listener2).toHaveBeenCalled();
expect(listener3).toHaveBeenCalled();
```

Reply:
I've added this to all 3 tests